core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -291,6 +293,10 @@ import scala.util.control.NonFatal
       commitMaps = offsets :: commitMaps
       commitSenders = commitSenders :+ sender()
 
+    case CommitWithoutReply(offsets) =>
+      // prepending, as later received offsets most likely are higher
+      commitMaps = offsets :: commitMaps
+
```

Comment:
I think it's great to add this alternative.

> It is not obvious why one would backpressure reading from a consumer when commits are not acknowledged by Kafka at the same rate. 

One thought is that if reading more messages is not slowed down by commits the gap between processed messages and actually committed messages might become big, maybe even grow unbounded if committing is slower than consuming.

Can that happen? What happens in `poll` when a commit is slow? It will block the actor until the commit has been confirmed by the Kafka server and then invoking the callback on the same thread (the actor thread that invoked poll)?

That would slow down the reads, since those are from the same poll.

Also, new commits for messages in flight will not be buffered in the actor but will be pending in the mailbox of the actor. Then all of them will be processed here by the actor before the next poll tick (since the poll tick is placed last in mailbox). Meaning that we can collapse all commits for a topic/partition into the latest and only emit that in next poll cycle. Maybe that is already how it works?

If that is how it works then my thought is probably less of a problem, but I suggest that we simulate the scenario to be sure.

Reply:
> One thought is that if reading more messages is not slowed down by commits the gap between processed messages and actually committed messages might become big, maybe even grow unbounded if committing is slower than consuming.

You're right: We need to be sure how fetching new records and committing interacts during `poll`. Hopefully @seglo 	has some insights here.

> Meaning that we can collapse all commits for a topic/partition into the latest and only emit that in next poll cycle. Maybe that is already how it works?

Yes, the offsets are aggregated nowadays to send just the latest offsets before every poll:

https://github.com/akka/alpakka-kafka/blob/42804d0b0f33197a8d832a957f380de2f1597040/core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala#L186-L195

Reply:
Thanks for clarifying, sounds promising